plugin/pulpcore/plugin/stages/association_stages.py

```diff
@@ -0,0 +1,158 @@
+import asyncio
+from collections import defaultdict
+
+from django.db.models import Q
+
+from pulpcore.plugin.models import ProgressBar
+
+
+class content_unit_association:
+    """
+    An object containing a Stages API stage that associates content units with `new_version`.
+
+    The actual stage is the :meth:`~pulpcore.plugin.stages.content_unit_association.stage`
+    which can be used as follows:
+
+    >>> content_unit_association(my_new_version).stage  # This is the real stage
+
+    This stage stores all content unit types and unit keys in memory before running. This is done to
+    compute the units already associated but not received from `in_q`. These units are passed via
+    `out_q` to the next stage as a :class:`django.db.models.query.QuerySet`.
+
+    This stage creates a ProgressBar named 'Associating Content' that counts the number of units
+    associated. Since it's a stream the total count isn't known until it's finished.
+
+    Args:
+        new_version (:class:`~pulpcore.plugin.models.RepositoryVersion`): The repo version this
+            stage associates content with.
+
+    Returns:
+        An object containing the content_unit_association stage to be included in a pipeline.
+    """
+
+    def __init__(self, new_version):
+        self.new_version = new_version
+        self.unit_keys_by_type = defaultdict(set)
+        for unit in self.new_version.content.all():
+            unit = unit.cast()
+            self.unit_keys_by_type[type(unit)].add(unit.natural_key())
+
+    async def stage(self, in_q, out_q):
+        """
+        For each Content Unit associate it with the repository version
+
+        Args:
+            in_q (:class:`asyncio.Queue`): Each item is a
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` with saved `content` that needs
+                to be associated.
+            out_q (:class:`asyncio.Queue`): Each item is a :class:`django.db.models.query.QuerySet`
+                of :class:`~pulpcore.plugin.models.Content` subclass that are already associated but
+                not included in the stream of items from `in_q`. One
+                :class:`django.db.models.query.QuerySet` is put for each
+                :class:`~pulpcore.plugin.models.Content` type.
+        """
+        with ProgressBar(message='Associating Content') as pb:
+            declarative_content_list = []
+            shutdown = False
+            while True:
+                try:
+                    declarative_content = in_q.get_nowait()
+                except asyncio.QueueEmpty:
+                    if not declarative_content_list and not shutdown:
+                        declarative_content = await in_q.get()
+                        declarative_content_list.append(declarative_content)
+                        continue
+                else:
+                    declarative_content_list.append(declarative_content)
+                    continue
```

Comment:
Same concern here ^ about batching algorithm.

Reply:
I should outline the performance measurement approach I thought up. We really need data to know if we need to make changes.